### PR TITLE
Metadata

### PR DESCRIPTION
**Pull request by yourheropaul**
This is an implementation of the metadata layer including:

- Repository-level metadata interface
- Git implementation of low level meta data accessors
- Interfaces for high level metadata

...

### CODE REVIEW

File: repository/repository_git_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+	require.True(t, repo.IsInitialised())
+}
+
+func Test_AnInitialisedGitRepositoryCanHashObjects(t *testing.T) {
+
+	_, repo, shutdown := initialisedGitRepoDir(t)
+	defer shutdown()
+
+	for cycle, test := range []struct {
```
Comment (adrianduke) [inline repository/repository_git_test.go]: Maybe use subtest? https://blog.golang.org/subtests

File: repository/repository_git.go [hunk 2]
```diff
@@ -37,36 +46,42 @@ type GitCmdErr struct {
...
+			err.ExitCode,
+			strings.Join(err.Args, " "),
+		)
 	}
 
 	return err.Stderr
 }
 
-type Git struct {
+type repositoryGit struct {
```
Comment (adrianduke) [inline repository/repository_git.go]: Is this a hangover, or did you rename it?

File: metadata/entry_reader.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import "time"
+
+// ID represents the address or other ID of the metadata in the repo
+type ID string
+
+// An EntryReader reads a metadata enty
+type EntryReader interface {
+	Id() ID
+	Created() time.Time
+	Data() interface{}
```
Comment (adrianduke) [inline metadata/entry_reader.go]: Is it worth making this Reader compatible? `Read(p []byte) (n int, err error)`, or is it too high level?

File: metadata/entry_writer.go [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package metadata
+
+// An EntryWriter can update a metadata entry
+type EntryWriter interface {
+	SetData(interface{}) error
```
Comment (adrianduke) [inline metadata/entry_writer.go]: Same here, Is it worth making this Writer compatible? `Write(p []byte) (n int, err error)`, or is it too high level?

Maybe in either case methods should be `Read` and `Write` respectively.

File: metadata/writer.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package metadata
+
+// Writer allows the creation (via appending) and deletion of metadata
+type Writer interface {
```
Comment (adrianduke) [inline metadata/writer.go]: In the `config` package we used `Storer`, same here or do you think they should be different?

File: repository/repository_git.go [hunk 6]
```diff
@@ -90,28 +105,57 @@ func (repo *Git) All() (map[string]string, error) {
...
+		return "", err
+	}
+
+	return repo.runGitCommand("notes", "--ref", notesRef, "show", id)
+}
+
+func (repo *repositoryGit) SetMetadata(key, value string) error {
+
+	id, err := repo.objectID(key)
+
```
Comment (adrianduke) [inline repository/repository_git.go]: Lots of spaces... What did we agree on? Is this and the next line not related?

File: repository/repository_git_test.go (hunk 1 above)
Comment (yourheropaul) [reply repository/repository_git_test.go]: good idea!

File: repository/repository_git.go (hunk 2 above)
Comment (yourheropaul) [reply repository/repository_git.go]: I renamed it to match the file name

File: repository/repository_git.go (hunk 6 above)
Comment (yourheropaul) [reply repository/repository_git.go]: Good question, I don't remember. I'll make it consistent

File: repository/repository_git.go (hunk 2 above)
Comment (yourheropaul) [reply repository/repository_git.go]: That's how it is, right? I just changed the variable name to be consistent with the file name.

File: metadata/entry_reader.go (hunk 3 above)
Comment (yourheropaul) [reply metadata/entry_reader.go]: You're right, it'd badly named. I will make it `Read` and `Write`.

File: metadata/writer.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package metadata
```
Comment (adrianduke) [inline metadata/writer.go]: Looks like this file needs to be removed?

File: repository/repository_git.go (hunk 2 above)
Comment (adrianduke) [reply repository/repository_git.go]: I guess what I was getting at is you'd use `<interface>_<implementation>.go` as the file name and `type <implementation> struct` as the type name. My thinking is that the package is already called `repository`, seems a bit redundant.

File: repository/repository_git.go [hunk 8]
```diff
@@ -156,40 +199,53 @@ func (repo *Git) runGitCommandRaw(args ...string) (string, string, int, error) {
...
 		args = append(args, "--system")
 	}
 
 	return strings.Join(args, " ")
 }
 
-func (repo *Git) configWriteScope() string {
-	switch repo.ConfigWriteScope {
-	case ScopeLocal:
+func (repo *repositoryGit) configWriteScopeArgs() string {
```
Comment (adrianduke) [inline repository/repository_git.go]: Should be `configWriteScopeArg` -- only returns 1 arg, not multiple like `configReadScopeArgs`

File: repository/repository_git.go (hunk 2 above)
Comment (yourheropaul) [reply repository/repository_git.go]: it's not the package name, it's the interface -- which is `Repository` as well.

**Pull request merged**